File: languages/go/entity/identity/credential_core_builder.go
package identity

import (
	"context"
	"errors"
	"time"

	"github.com/bloock/bloock-sdk-go/v2/internal/bridge"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
)

// CredentialCoreBuilder helps construct credentials by specifying various attributes.
type CredentialCoreBuilder struct {
	schemaId   string
	issuerDid  string
	holderDid  string
	expiration int64
	version    int32
	key        *proto.Key
	configData *proto.ConfigData

	stringAttribute   []StringAttribute
	integerAttribute  []IntegerAttribute
	decimalAttribute  []DecimalAttribute
	booleanAttribute  []BooleanAttribute
	dateAttribute     []DateAttribute
	datetimeAttribute []DatetimeAttribute
}

// NewCredentialCoreBuilder creates a new CredentialCoreBuilder instance with the specified parameters.
func NewCredentialCoreBuilder(issuer Issuer, schemaId, holderDid string, expiration int64, version int32, configData *proto.ConfigData) CredentialCoreBuilder {
	return CredentialCoreBuilder{
		schemaId:          schemaId,
		issuerDid:         issuer.Did.Did,
		holderDid:         holderDid,
		expiration:        expiration,
		version:           version,
		configData:        configData,
		key:               issuer.Key.ToProto(),
		stringAttribute:   []StringAttribute{},
		integerAttribute:  []IntegerAttribute{},
		decimalAttribute:  []DecimalAttribute{},
		booleanAttribute:  []BooleanAttribute{},
		dateAttribute:     []DateAttribute{},
		datetimeAttribute: []DatetimeAttribute{},
	}
}

// WithStringAttribute adds a string attribute to the CredentialCoreBuilder.
func (c CredentialCoreBuilder) WithStringAttribute(key string, value string) CredentialCoreBuilder {
	c.stringAttribute = append(c.stringAttribute, NewStringAttribute(key, value))
	return c
}

// WithIntegerAttribute adds an integer attribute to the CredentialCoreBuilder.
func (c CredentialCoreBuilder) WithIntegerAttribute(key string, value int64) CredentialCoreBuilder {
	c.integerAttribute = append(c.integerAttribute, NewIntegerAttribute(key, value))
	return c
}

// WithDecimalAttribute adds a decimal attribute to the CredentialCoreBuilder.
func (c CredentialCoreBuilder) WithDecimalAttribute(key string, value float64) CredentialCoreBuilder {
	c.decimalAttribute = append(c.decimalAttribute, NewDecimalAttribute(key, value))
	return c
}

// WithBooleanAttribute adds a boolean attribute to the CredentialCoreBuilder.
func (c CredentialCoreBuilder) WithBooleanAttribute(key string, value bool) CredentialCoreBuilder {
	c.booleanAttribute = append(c.booleanAttribute, NewBooleanAttribute(key, value))
	return c
}

// WithDateAttribute adds a date attribute to the CredentialCoreBuilder.
func (c CredentialCoreBuilder) WithDateAttribute(key string, value time.Time) CredentialCoreBuilder {
	c.dateAttribute = append(c.dateAttribute, NewDateAttribute(key, value))
	return c
}

// WithDatetimeAttribute adds a datetime attribute to the CredentialCoreBuilder.
func (c CredentialCoreBuilder) WithDatetimeAttribute(key string, value time.Time) CredentialCoreBuilder {
	c.datetimeAttribute = append(c.datetimeAttribute, NewDatetimeAttribute(key, value))
	return c
}

// Build creates and returns a Credential using the specified attributes.
func (c CredentialCoreBuilder) Build() (CredentialReceipt, error) {
	bridge := bridge.NewBloockBridge()

	stringAttributes := make([]*proto.StringAttribute, len(c.stringAttribute))
	for i, b := range c.stringAttribute {
		stringAttributes[i] = b.ToProto()
	}

	integerAttributes := make([]*proto.IntegerAttribute, len(c.integerAttribute))
	for i, b := range c.integerAttribute {
		integerAttributes[i] = b.ToProto()
	}

	decimalAttributes := make([]*proto.DecimalAttribute, len(c.decimalAttribute))
	for i, b := range c.decimalAttribute {
		decimalAttributes[i] = b.ToProto()
	}

	booleanAttributes := make([]*proto.BooleanAttribute, len(c.booleanAttribute))
	for i, b := range c.booleanAttribute {
		booleanAttributes[i] = b.ToProto()
	}

	dateAttributes := make([]*proto.DateAttribute, len(c.dateAttribute))
	for i, b := range c.dateAttribute {
		dateAttributes[i] = b.ToProto()
	}

	datetimeAttributes := make([]*proto.DateTimeAttribute, len(c.datetimeAttribute))
	for i, b := range c.datetimeAttribute {
		datetimeAttributes[i] = b.ToProto()
	}

	req := proto.CreateCoreCredentialRequest{
		SchemaId:           c.schemaId,
		IssuerDid:          c.issuerDid,
		HolderDid:          c.holderDid,
		Expiration:         c.expiration,
		Version:            &c.version,
		Key:                c.key,
		ConfigData:         c.configData,
		StringAttributes:   stringAttributes,
		IntegerAttributes:  integerAttributes,
		DecimalAttributes:  decimalAttributes,
		BooleanAttributes:  booleanAttributes,
		DateAttributes:     dateAttributes,
		DatetimeAttributes: datetimeAttributes,
	}

	res, err := bridge.IdentityCore().CreateCoreCredential(context.Background(), &req)
	if err != nil {
		return CredentialReceipt{}, err
	}

	if res.Error != nil {
		return CredentialReceipt{}, errors.New(res.Error.Message)
	}

	return NewCredentialReceiptFromProto(res.GetCredentialReceipt()), nil
}

File: languages/js/src/entity/identity/credential_core_builder.ts
import { BloockBridge } from "../../bridge/bridge";
import { ConfigData } from "../../bridge/proto/config";
import { CreateCoreCredentialRequest } from "../../bridge/proto/identity_core";
import {
  BooleanAttribute as BooleanAttributeProto,
  DateAttribute as DateAttributeProto,
  DateTimeAttribute as DateTimeAttributeProto,
  DecimalAttribute as DecimalAttributeProto,
  IntegerAttribute as IntegerAttributeProto,
  StringAttribute as StringAttributeProto
} from "../../bridge/proto/identity_entities";
import { Key as KeyProto } from "../../bridge/proto/keys_entities";
import { BooleanAttribute } from "./boolean_attribute";
import { CredentialReceipt } from "./credential_receipt";
import { DateAttribute } from "./date_attribute";
import { DateTimeAttribute } from "./datetime_attribute";
import { DecimalAttribute } from "./decimal_attribute";
import { IntegerAttribute } from "./integer_attribute";
import { Issuer } from "./issuer";
import { StringAttribute } from "./string_attribute";

/**
 * Helps construct credentials by specifying various attributes.
 */
export class CredentialCoreBuilder {
  schemaId: string;
  issuerDid: string;
  holderDid: string;
  expiration: number;
  version: number;
  key?: KeyProto;
  configData: ConfigData;

  stringAttributes: StringAttributeProto[];
  integerAttributes: IntegerAttributeProto[];
  decimalAttributes: DecimalAttributeProto[];
  booleanAttributes: BooleanAttributeProto[];
  dateAttributes: DateAttributeProto[];
  dateTimeAttributes: DateTimeAttributeProto[];

  /**
   * Creates a new CredentialCoreBuilder instance with the specified parameters.
   * @param schemaId
   * @param issuerDid
   * @param holderDid
   * @param expiration
   * @param version
   * @param configData
   */
  constructor(
    issuer: Issuer,
    schemaId: string,
    holderDid: string,
    expiration: number,
    version: number,
    configData: ConfigData
  ) {
    this.schemaId = schemaId;
    this.issuerDid = issuer.did.did;
    this.holderDid = holderDid;
    this.expiration = expiration;
    this.version = version;
    this.configData = configData;

    this.key = issuer.key.toProto();
    this.stringAttributes = [];
    this.integerAttributes = [];
    this.decimalAttributes = [];
    this.booleanAttributes = [];
    this.dateAttributes = [];
    this.dateTimeAttributes = [];
  }

  /**
   * Adds a string attribute to the CredentialCoreBuilder.
   * @param key
   * @param value
   * @returns
   */
  public withStringAttribute(
    key: string,
    value: string
  ): CredentialCoreBuilder {
    this.stringAttributes.push(new StringAttribute(key, value));
    return this;
  }

  /**
   * Adds an integer attribute to the CredentialCoreBuilder.
   * @param key
   * @param value
   * @returns
   */
  public withIntegerAttribute(
    key: string,
    value: number
  ): CredentialCoreBuilder {
    this.integerAttributes.push(new IntegerAttribute(key, value));
    return this;
  }

  /**
   * Adds a decimal attribute to the CredentialCoreBuilder.
   * @param key
   * @param value
   * @returns
   */
  public withDecimalAttribute(
    key: string,
    value: number
  ): CredentialCoreBuilder {
    this.decimalAttributes.push(new DecimalAttribute(key, value));
    return this;
  }

  /**
   * Adds a boolean attribute to the CredentialCoreBuilder.
   * @param key
   * @param value
   * @returns
   */
  public withBooleanAttribute(
    key: string,
    value: boolean
  ): CredentialCoreBuilder {
    this.booleanAttributes.push(new BooleanAttribute(key, value));
    return this;
  }

  /**
   * Adds a date attribute to the CredentialCoreBuilder.
   * @param key
   * @param value
   * @returns
   */
  public withDateAttribute(key: string, value: Date): CredentialCoreBuilder {
    this.dateAttributes.push(new DateAttribute(key, value).toProto());
    return this;
  }

  /**
   * Adds a datetime attribute to the CredentialCoreBuilder.
   * @param key
   * @param value
   * @returns
   */
  public withDateTimeAttribute(
    key: string,
    value: Date
  ): CredentialCoreBuilder {
    this.dateTimeAttributes.push(new DateTimeAttribute(key, value).toProto());
    return this;
  }

  /**
   * Creates and returns a Credential using the specified attributes.
   * @returns
   */
  async build(): Promise<CredentialReceipt> {
    const bridge = new BloockBridge();

    const req = CreateCoreCredentialRequest.fromPartial({
      configData: this.configData,
      schemaId: this.schemaId,
      issuerDid: this.issuerDid,
      holderDid: this.holderDid,
      expiration: this.expiration,
      version: this.version,
      key: this.key,
      stringAttributes: this.stringAttributes,
      integerAttributes: this.integerAttributes,
      decimalAttributes: this.decimalAttributes,
      booleanAttributes: this.booleanAttributes,
      dateAttributes: this.dateAttributes,
      datetimeAttributes: this.dateTimeAttributes
    });

    return bridge
      .getIdentityCore()
      .CreateCoreCredential(req)
      .then(res => {
        if (res.error) {
          throw res.error;
        }
        return CredentialReceipt.fromProto(res.credentialReceipt!);
      });
  }
}
